packages/core/src/sync-historical/index.ts

```diff
@@ -65,9 +74,19 @@ export const createHistoricalSync = async (
    * Note: All entries are deleted at the end of each call to `sync()`.
    */
   const transactionsCache = new Set<Hash>();
-
-  // const logMetadata = new Map<LogFilter, { range: number }>();
-
+  /**
+   * Data about the range passed to "eth_getLogs" for all log
+   *  filters and log factories.
+   */
+  const getLogsRequestMetadata = new Map<
+    LogFilter | LogFactory,
+    {
+      /** Estimate optimal range to use for "eth_getLogs" requests */
+      range: number;
+      /** `true` if an exact range has been suggested by the error message */
+      isSuggestedByError: boolean;
```

Comment:
Another idea would be `estimatedRange` and `confirmedRange`, where `confirmedRange` only gets set when we are certain of a hard cap. Then always prefer `confirmedRange` and fall back to `estimatedRange` if not present.